server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java

```diff
@@ -0,0 +1,142 @@
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(
+            Signature.scalar(
+                NAME,
+                DataTypes.STRING.getTypeSignature(),
+                DataTypes.UNTYPED_OBJECT.getTypeSignature()
+            ),
+            ParseURLFunction::new
+        );
+    }
+
+    private final Signature signature;
+    private final BoundSignature boundSignature;
+
+    public ParseURLFunction(Signature signature, BoundSignature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    public Signature signature() {
+        return signature;
+    }
+
+    @Override
+    public BoundSignature boundSignature() {
+        return boundSignature;
+    }
+
+    @Override
+    @SafeVarargs
+    public final Object evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<String>... args) {
+        String url = args[0].value();
+        if (url == null) {
+            return null;
+        }
+        return parseURL(url);
+    }
+
+    private final Object parseURL(String urlText) {
+        final Map<String, Object> urlMap = new LinkedHashMap<>();
+
+        URL url = null;
+
+        try {
+            url = new URL(urlText);
+        } catch (MalformedURLException e1) {
+            throw new IllegalArgumentException(String.format(Locale.ENGLISH,
+                                                            "unable to parse url %s",
+                                                             urlText));
+        }
+
+        urlMap.put("scheme", url.getProtocol());
+        urlMap.put("userinfo", decodeText(url.getUserInfo()));
+        urlMap.put("hostname", url.getHost());
+        urlMap.put("port", url.getPort() == -1 ? null : url.getPort());
+        urlMap.put("path", decodeText(url.getPath()));
+        urlMap.put("query", decodeText(url.getQuery()));
+        urlMap.put("parameters", parseQuery(url.getQuery()));
+        urlMap.put("fragment", decodeText(url.getRef()));
+
+        return urlMap;
+    }
+
+    private final Map<String,List<String>> parseQuery(String query) {
+        if (Strings.isNullOrEmpty(query)) {
+            return null;
+        }
+
+        Map<String,List<String>> queryMap = new HashMap<String,List<String>>();
+        Arrays.stream(query.split("&(?!amp)"))
+                     .forEach((String param) -> parseQueryParameters(queryMap, param));
+
+        return queryMap;
+    }
+
+    private final void parseQueryParameters(Map<String,List<String>> map, String params) {
+        final int idx = params.indexOf("=");
+        final String key = idx > 0 ? decodeText(params.substring(0, idx)) : decodeText(params);
+        final String value = idx > 0 && params.length() > idx + 1 ? decodeText(params.substring(idx + 1)) : null;
+        if (!map.containsKey(key)) {
+            map.put(key, new ArrayList<String>());
+        }
+        map.get(key).add(value);
```

Comment:
I think map.computeIfAbsent(key, k -> new ArrayList<>()).add(value); would work here - less verbose and saves one lookup if key exits.

Regarding `ArrayList` - maybe can be `Set` depending on what we want. For query string like `key=val1&another_key=val&key=val1`, we end up having key-> (val, val). It's fine but we could drop the duplicates. If users want original query string, they can use "query" field but parse_url("someurl")[parameters] can have unique vals. No strong opinion here.

Reply:
changed to `computeIfAbsent`

I would prefer to use an ArrayList to be able to recreate the original URL, as the `query` string is decoded, which is not always reversible